flocker/node/agents/testtools/_blockdevice.py

```diff
@@ -869,3 +882,289 @@ def setUp(self):
+        config = yaml.safe_load(config_file.read())
+
+    section = config.get(platform_name)
+    if section is None:
+        raise InvalidConfig(
+            "The requested cloud platform "
+            "was not found in the configuration file. "
+            "Platform: %s, "
+            "Configuration File: %s" % (platform_name, config_file_path)
+        )
+
+    provider_name = section.get('provider', platform_name)
+    try:
+        provider_environment = ProviderType.lookupByName(provider_name)
+    except ValueError:
+        raise InvalidConfig(
+            "Unsupported provider. "
+            "Supplied provider: %s, "
+            "Available providers: %s" % (
+                provider_name,
+                ', '.join(p.name for p in ProviderType.iterconstants())
+            )
+        )
+
+    if provider_environment != provider_type:
+        raise InvalidConfig(
+            "The requested cloud provider (%s) is not the provider running "
+            "the tests (%s)." % (provider_type.name, provider_environment.name)
+        )
+
+    # XXX - If provider type is Rackspace, we add the auth_plugin config
+    # here.  The CI system should be configured to provide this
+    # parameter as part of the config, to match the documented way of
+    # configuring Rackspace.
+    if provider_environment == ProviderType.rackspace:
+        section['auth_plugin'] = 'rackspace'
+
+    return section
+
+
+def get_openstack_region_for_test():
+    """
+    Return a default Openstack region for testing.
+
+    The default region comes from an environment variable.  Keystone
+    uses case-sensitive regions, so ensure region is uppercase.
+    """
+    region = environ.get('FLOCKER_FUNCTIONAL_TEST_OPENSTACK_REGION')
+    if region is not None:
+        region = region.upper()
+    return region
+
+
+def _openstack(cluster_id, config):
+    """
+    Create an IBlockDeviceAPI provider configured to use the Openstack
+    region where the server that is running this code is running.
+
+    :param config: Any additional configuration (possibly provider-specific)
+        necessary to authenticate a keystone session.
+    :return: A CinderBlockDeviceAPI instance.
+    """
+    region = get_openstack_region_for_test()
+    return cinder_from_configuration(region, cluster_id, **config)
+
+
+def get_ec2_client_for_test(config):
+    """
+    Get a simple EC2 client, configured for the test region.
+    """
+
+    # We just get the credentials from the config file.
+    # We ignore the region specified in acceptance test configuration,
+    # and instead get the region from the zone of the host.
+    zone = environ['FLOCKER_FUNCTIONAL_TEST_AWS_AVAILABILITY_ZONE']
+    # The region is the zone, without the trailing [abc].
+    region = zone[:-1]
+    return ec2_client(
+        region=region,
+        zone=zone,
+        access_key_id=config['access_key'],
+        secret_access_key=config['secret_access_token']
+    )
```

Comment:
This AWS specific function shouldn't be here, but I'm going to remove it in a followup branch so thought it'd be easier to move everything to a single place for ease of review.